File: javascript/nerkh-gold-price-fetcher.js
import axios from 'axios';
import chalk from 'chalk';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

// Configuration
const CONFIG = {
    apiUrl: process.env.API_URL,
    apiKey: process.env.API_KEY
};

// Gold price types to fetch
const GOLD_TYPES = {
    BAHAR: 'Bahare Azadi Coin',
    EMAMI: 'Emami Coin',
    NIM: 'Half Coin',
    ROB: 'Quarter Coin',
    GOLD_18: '18K Gold',
    GOLD_24: '24K Gold',
};

class NerkhGoldPriceFetcher {
    constructor() {
        this.prices = {};
        this.axiosInstance = axios.create({
            baseURL: CONFIG.apiUrl,
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${CONFIG.apiKey}`
            }
        });
    }

    async fetchPrices() {
        try {
            const response = await this.axiosInstance.get('');
            const data = response.data;
            this.prices = this.processPrices(data);
            this.displayPrices();
            return this.prices;
        } catch (error) {
            throw error;
        }
    }

    displayPrices() {
        Object.entries(this.prices).forEach(([type, data]) => {
            console.log(chalk.bold.cyan(`${type}:`));
            console.log(`  Current Price: ${chalk.green(data.price)}`);
            console.log(`  1-hour Range: ${chalk.yellow(data.min_1h)} - ${chalk.yellow(data.max_1h)} (${this.formatChange(data.change_1h)})`);
            console.log(`  12-hour Range: ${chalk.magenta(data.min_12h)} - ${chalk.magenta(data.max_12h)} (${this.formatChange(data.change_12h)})`);
            console.log(`  Last Update: ${chalk.gray(data.timestamp)}`);
            console.log('---');
        });
    }

    processPrices(data) {
        // Get the prices from the nested data structure
        const prices = data.data.prices;
        
        // Define the mapping between our types and API types
        const priceTypes = {
            [GOLD_TYPES.BAHAR]: 'SEKE_BAHAR',
            [GOLD_TYPES.EMAMI]: 'SEKE_EMAMI',
            [GOLD_TYPES.NIM]: 'SEKE_NIM',
            [GOLD_TYPES.ROB]: 'SEKE_ROB',
            [GOLD_TYPES.GOLD_18]: 'GOLD18K',
            [GOLD_TYPES.GOLD_24]: 'GOLD24K'
        };

        // Process all gold types in a loop
        return Object.entries(priceTypes).reduce((result, [type, apiType]) => {
            result[type] = {
                price: this.formatPrice(prices[apiType].current),
                min_1h: this.formatPrice(prices[apiType].min['1hour']),
                max_1h: this.formatPrice(prices[apiType].max['1hour']),
                min_12h: this.formatPrice(prices[apiType].min['12hour']),
                max_12h: this.formatPrice(prices[apiType].max['12hour']),
                timestamp: prices[apiType].update,
                change_1h: this.calculateChange(
                    parseInt(prices[apiType].current),
                    parseInt(prices[apiType].min['1hour']),
                    parseInt(prices[apiType].max['1hour'])
                ),
                change_12h: this.calculateChange(
                    parseInt(prices[apiType].current),
                    parseInt(prices[apiType].min['12hour']),
                    parseInt(prices[apiType].max['12hour'])
                )
            };
            return result;
        }, {});
    }

    calculateChange(current, min, max) {
        const range = max - min;
        if (range === 0) return 0;
        const position = current - min;
        return (position / range) * 100;
    }

    formatPrice(price) {
        // Format price with commas
        return new Intl.NumberFormat('en-US').format(price);
    }

    formatChange(change) {
        const color = change >= 50 ? chalk.green : chalk.red;
        return color(`${change.toFixed(1)}%`);
    }
}

// Export for use in other modules if needed
export default NerkhGoldPriceFetcher;

// Create an instance and run the fetcher
const fetcher = new NerkhGoldPriceFetcher();
fetcher.fetchPrices()
    .then(() => {
        console.log('Gold prices fetched successfully');
    })
    .catch(error => {
        console.error('Error fetching gold prices:', error.message);
        process.exit(1);
    }); 

File: python/nerkh_gold_price_fetcher.py
import os
import requests
from datetime import datetime
from dotenv import load_dotenv
from colorama import init, Fore, Style

# Initialize colorama
init()

# Load environment variables
load_dotenv()

# Configuration
CONFIG = {
    'api_url': os.getenv('API_URL'),
    'api_key': os.getenv('API_KEY')
}

# Gold price types to fetch
GOLD_TYPES = {
    'BAHAR': 'Bahare Azadi Coin',
    'EMAMI': 'Emami Coin',
    'NIM': 'Half Coin',
    'ROB': 'Quarter Coin',
    'GOLD_18': '18K Gold',
    'GOLD_24': '24K Gold'
}

class NerkhGoldPriceFetcher:
    def __init__(self):
        self.prices = {}
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {CONFIG["api_key"]}'
        })

    def fetch_prices(self):
        try:
            response = self.session.get(CONFIG['api_url'])
            response.raise_for_status()
            data = response.json()
            self.prices = self.process_prices(data)
            self.display_prices()
            return self.prices
        except requests.exceptions.RequestException as e:
            print(f"{Fore.RED}Error fetching gold prices: {str(e)}{Style.RESET_ALL}")
            raise

    def display_prices(self):
        for type_name, data in self.prices.items():
            print(f"{Fore.CYAN}{Style.BRIGHT}{type_name}:{Style.RESET_ALL}")
            print(f"  Current Price: {Fore.GREEN}{data['price']}{Style.RESET_ALL}")
            print(f"  1-hour Range: {Fore.YELLOW}{data['min_1h']} - {data['max_1h']} ({self.format_change(data['change_1h'])}){Style.RESET_ALL}")
            print(f"  12-hour Range: {Fore.MAGENTA}{data['min_12h']} - {data['max_12h']} ({self.format_change(data['change_12h'])}){Style.RESET_ALL}")
            print(f"  Last Update: {Fore.LIGHTBLACK_EX}{data['timestamp']}{Style.RESET_ALL}")
            print('---')

    def process_prices(self, data):
        prices = data['data']['prices']
        
        price_types = {
            GOLD_TYPES['BAHAR']: 'SEKE_BAHAR',
            GOLD_TYPES['EMAMI']: 'SEKE_EMAMI',
            GOLD_TYPES['NIM']: 'SEKE_NIM',
            GOLD_TYPES['ROB']: 'SEKE_ROB',
            GOLD_TYPES['GOLD_18']: 'GOLD18K',
            GOLD_TYPES['GOLD_24']: 'GOLD24K'
        }

        result = {}
        for type_name, api_type in price_types.items():
            result[type_name] = {
                'price': self.format_price(prices[api_type]['current']),
                'min_1h': self.format_price(prices[api_type]['min']['1hour']),
                'max_1h': self.format_price(prices[api_type]['max']['1hour']),
                'min_12h': self.format_price(prices[api_type]['min']['12hour']),
                'max_12h': self.format_price(prices[api_type]['max']['12hour']),
                'timestamp': prices[api_type]['update'],
                'change_1h': self.calculate_change(
                    int(prices[api_type]['current']),
                    int(prices[api_type]['min']['1hour']),
                    int(prices[api_type]['max']['1hour'])
                ),
                'change_12h': self.calculate_change(
                    int(prices[api_type]['current']),
                    int(prices[api_type]['min']['12hour']),
                    int(prices[api_type]['max']['12hour'])
                )
            }
        return result

    def calculate_change(self, current, min_price, max_price):
        range_price = max_price - min_price
        if range_price == 0:
            return 0
        position = current - min_price
        return (position / range_price) * 100

    def format_price(self, price):
        return f"{int(price):,}"

    def format_change(self, change):
        color = Fore.GREEN if change >= 50 else Fore.RED
        return f"{color}{change:.1f}%{Style.RESET_ALL}"

if __name__ == "__main__":
    fetcher = NerkhGoldPriceFetcher()
    try:
        fetcher.fetch_prices()
        print(f"{Fore.GREEN}Gold prices fetched successfully{Style.RESET_ALL}")
    except Exception as e:
        print(f"{Fore.RED}Error: {str(e)}{Style.RESET_ALL}")
        exit(1) 